include/particles/particle_functions.tcc

```diff
@@ -0,0 +1,121 @@
+// Compute mass of particle
+namespace mpm {
+namespace particle {
+
+// Compute particle mass
+template <unsigned Tdim>
+void compute_mass(std::shared_ptr<mpm::ParticleBase<Tdim>> particle) noexcept {
+  // Check if particle volume is set and material ptr is valid
+  assert(particle->volume() != std::numeric_limits<double>::max() &&
+         particle->material() != nullptr);
+
+  // Mass = volume of particle * mass_density
+  particle->update_scalar_property(
+      mpm::properties::Scalar::MassDensity, false,
+      particle->material()->template property<double>(std::string("density")));
+
+  // Update particle mass
+  particle->update_scalar_property(
+      mpm::properties::Scalar::Mass, false,
+      particle->volume() * particle->mass_density());
+}
+
+// Compute volume of the particle
+template <unsigned Tdim>
+void compute_volume(
+    std::shared_ptr<mpm::ParticleBase<Tdim>> particle) noexcept {
+  // Check if particle has a valid cell ptr
+  assert(particle->cell_ptr());
+
+  // Volume of the cell / # of particles
+  particle->assign_volume(particle->cell()->volume() /
+                          particle->cell()->nparticles());
+}
+
+// Update volume based on the central strain rate
+template <unsigned Tdim>
+void update_volume(std::shared_ptr<mpm::ParticleBase<Tdim>> particle) noexcept {
+  // Check if particle has a valid cell ptr and a valid volume
+  assert(particle->cell_ptr() &&
+         particle->volume() != std::numeric_limits<double>::max());
+
+  // Compute at centroid
+  // Strain rate for reduced integration
+  particle->update_scalar_property(
+      mpm::properties::Scalar::Volume, false,
+      (particle->volume() * (1. + particle->dvolumetric_strain())));
+  particle->update_scalar_property(
+      mpm::properties::Scalar::MassDensity, false,
+      (particle->mass_density() / (1. + particle->dvolumetric_strain())));
+}
+
+//! Map particle mass and momentum to nodes
+template <unsigned Tdim>
+void map_mass_momentum_to_nodes(
+    std::shared_ptr<mpm::ParticleBase<Tdim>> particle) noexcept {
+  // Check if particle mass is set
+  assert(particle->mass() != std::numeric_limits<double>::max());
+
+  // Map mass and momentum to nodes
+  particle->map_scalar_property_nodes(mpm::properties::Scalar::Mass, true,
+                                      mpm::ParticlePhase::Solid);
```

Comment:
I think the `map_scalar_property_nodes` has a problem to be used for two phase. The liquid mass is called `LiquidMass` in the particle properties, but `Mass` at liquid_phase in the node properties. That's conflict.

Reply:
Use the overload function with input value as your `LiquidMass`
https://github.com/cb-geo/mpm/blob/164491fd7f5c7ef09536405aeb2ae0d5fd8947ea/include/particles/particle_base.tcc#L50-L53